File: lib/dedupe.js
import crypto from 'crypto';

/** Стабильный ключ URL okcdn (параметр r) */
export function urlDedupeKey(url) {
  if (!url) return '';
  const m = url.match(/[?&]r=([A-Za-z0-9_-]+)/);
  if (m) return `r:${m[1]}`;
  return url.split('&')[0].split('?')[0];
}

export function contentHash(data) {
  const buf = Buffer.isBuffer(data) ? data : Buffer.from(data);
  return crypto.createHash('sha256').update(buf).digest('hex').slice(0, 16);
}

export function dedupeStateFromManifest(manifest) {
  const urlKeys = new Set();
  const contentHashes = new Set();
  for (const it of manifest) {
    if (it.contentHash) contentHashes.add(it.contentHash);
    const u = it.originalUrl || it.url;
    if (u) urlKeys.add(urlDedupeKey(u));
  }
  return { urlKeys, contentHashes };
}

export function createBatchDedupe() {
  return { urlKeys: new Set(), contentHashes: new Set() };
}

/** null = не дубликат, иначе 'url' | 'hash' */
export function duplicateReason(state, batch, url, data) {
  const uk = url ? urlDedupeKey(url) : '';
  if (uk && (state.urlKeys.has(uk) || batch.urlKeys.has(uk))) return 'url';
  if (data?.length) {
    const h = contentHash(data);
    if (state.contentHashes.has(h) || batch.contentHashes.has(h)) return 'hash';
  }
  return null;
}

export function registerDedupe(state, batch, url, data) {
  const uk = url ? urlDedupeKey(url) : '';
  if (uk) {
    state.urlKeys.add(uk);
    batch.urlKeys.add(uk);
  }
  if (data?.length) {
    const h = contentHash(data);
    state.contentHashes.add(h);
    batch.contentHashes.add(h);
  }
  return data?.length ? contentHash(data) : undefined;
}

File: dedupe.py
"""Дедупликация фото по URL (okcdn r=) и хешу содержимого."""

from __future__ import annotations

import hashlib
import re
from typing import Any

R_PARAM_RE = re.compile(r"[?&]r=([A-Za-z0-9_-]+)")


def url_dedupe_key(url: str | None) -> str:
    if not url:
        return ""
    m = R_PARAM_RE.search(url)
    if m:
        return f"r:{m.group(1)}"
    return url.split("&")[0].split("?")[0]


def content_hash(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()[:16]


def dedupe_state_from_manifest(manifest: list[dict[str, Any]]) -> tuple[set[str], set[str]]:
    url_keys: set[str] = set()
    content_hashes: set[str] = set()
    for it in manifest:
        if it.get("contentHash"):
            content_hashes.add(it["contentHash"])
        u = it.get("originalUrl") or it.get("url")
        if u:
            url_keys.add(url_dedupe_key(u))
    return url_keys, content_hashes


def is_duplicate(
    url_keys: set[str],
    content_hashes: set[str],
    batch_url_keys: set[str],
    batch_hashes: set[str],
    url: str | None,
    data: bytes | None,
) -> bool:
    uk = url_dedupe_key(url) if url else ""
    if uk and (uk in url_keys or uk in batch_url_keys):
        return True
    if data:
        h = content_hash(data)
        if h in content_hashes or h in batch_hashes:
            return True
    return False


def register_dedupe(
    url_keys: set[str],
    content_hashes: set[str],
    batch_url_keys: set[str],
    batch_hashes: set[str],
    url: str | None,
    data: bytes | None,
) -> str | None:
    uk = url_dedupe_key(url) if url else ""
    if uk:
        url_keys.add(uk)
        batch_url_keys.add(uk)
    h = None
    if data:
        h = content_hash(data)
        content_hashes.add(h)
        batch_hashes.add(h)
    return h
